flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/OnlineKMeans.java

```diff
@@ -0,0 +1,437 @@
+        @Override
+        public DenseVector map(Row row) throws Exception {
+            return (DenseVector) row.getField(featuresCol);
+        }
+    }
+
+    // An operator that splits a global batch into evenly-sized local batches, and distributes them
+    // to downstream operator.
+    private static class GlobalBatchSplitter
+            implements FlatMapFunction<DenseVector[], DenseVector[]> {
+        private final int downStreamParallelism;
+
+        private GlobalBatchSplitter(int downStreamParallelism) {
+            this.downStreamParallelism = downStreamParallelism;
+        }
+
+        @Override
+        public void flatMap(DenseVector[] values, Collector<DenseVector[]> collector) {
+            // Calculate the batch sizes to be distributed on each subtask.
+            List<Integer> sizes = new ArrayList<>();
+            for (int i = 0; i < downStreamParallelism; i++) {
+                int start = i * values.length / downStreamParallelism;
+                int end = (i + 1) * values.length / downStreamParallelism;
+                sizes.add(end - start);
+            }
+
+            int offset = 0;
+            for (Integer size : sizes) {
+                collector.collect(Arrays.copyOfRange(values, offset, offset + size));
+                offset += size;
+            }
+        }
+    }
+
+    private static class GlobalBatchCreator
+            implements AllWindowFunction<DenseVector, DenseVector[], GlobalWindow> {
+        @Override
+        public void apply(
+                GlobalWindow timeWindow,
+                Iterable<DenseVector> iterable,
+                Collector<DenseVector[]> collector) {
+            List<DenseVector> points = IteratorUtils.toList(iterable.iterator());
+            collector.collect(points.toArray(new DenseVector[0]));
+        }
+    }
+
+    /**
+     * Sets the initial model data of the online training process with the provided model data
+     * table.
+     *
+     * <p>This would override the effect of previously invoked {@link #setInitialModelData(Table)}
+     * or {@link #setRandomCentroids(StreamTableEnvironment, int, int, double)}.
+     */
+    public OnlineKMeans setInitialModelData(Table initModelDataTable) {
+        this.initModelDataTable = initModelDataTable;
+        return this;
+    }
+
+    /**
+     * Sets the initial model data of the online training process with randomly created centroids.
+     *
+     * <p>This would override the effect of previously invoked {@link #setInitialModelData(Table)}
+     * or {@link #setRandomCentroids(StreamTableEnvironment, int, int, double)}.
+     *
+     * @param tEnv The stream table environment to create the centroids in.
+     * @param dim The dimension of the centroids to create.
+     * @param k The number of centroids to create.
+     * @param weight The weight of the centroids to create.
+     */
+    public OnlineKMeans setRandomCentroids(
+            StreamTableEnvironment tEnv, int dim, int k, double weight) {
```

Comment:
The introduction of `k` in `OnlineKMeans` might cause the following problem.

- `setK()` must be invoked before `setRandomCentroids()`.
- If users use `setInitialModelData`, then we need to check whether the initial model data would provide exactly `k` centroids.

Because of these problems, I am not sure whether we should add the K parameter on `OnlineKMeans`.

Reply:
Regarding the first issue, how about we just cache the variable and construct this table when fit() is invoked?

Regarding the second issue, whether or now we check initial model data seems like an implementation detail, which is probably less important than the API issue (e.g. whether we allow user to query the K via `getK()`).

Reply:
About the first issue, it means we still need to save `dim`, `weight` and `initMode` as metadata of this stage, and the most handy choice is to register them as WithParams interface. But we still want to expose `setRandomCentroids` as the API, instead of `setInitMode().setDim().setWeight()`, right? Does it mean we should still define these parameters, but not expose their get/set methods?

I agree with the second issue.

Reply:
Suppose all options can meet our use-case (e.g. allow user to get the information they need), then we need to choose the option that is more usable, simpler and easier to understand.

The issue with not exposing getK() is that `K` seems like an important attribute of the clustering algorithm and users will want to get this information. Maybe @zhipeng93 can comment on this.

The issue with using `setInitMode().setDim().setWeight() ` is that user needs to call much 3 methods instead of 1 method to set random centroids and understand that setDim() and setWeight() depends on initMode. This could make the algorithm harder to use. What do you think?

Reply:
I tried adding parameters of `DIM` and `INIT_WEIGHT` so as to save them with other metadata, but have not provided their getter/setter methods to still encourage users to use the `setRandomCentroids(dim, weight)` API. Please check if this is a proper design.

Reply:
According to offline discussion, I'll add a `KMeansUtils.generateRandomModelData()` method and remove the `setRandomCentroids` method from `OnlineKMeans`.